## Libraries

In [26]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Analyse exploratoire

In [27]:
df = pd.read_csv("vehicles.csv")

In [28]:
# Afficher les cinq premières lignes du df
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [29]:
# Afficher les cinq dernières lignes du df
df.tail()

id                                                url  \
426875  7301591192  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  7301591187  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  7301591147  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  7301591140  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  7301591129  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

         region                      region_url  price    year manufacturer  \
426875  wyoming  https://wyoming.craigslist.org  23590  2019.0       nissan   
426876  wyoming  https://wyoming.craigslist.org  30590  2020.0        volvo   
426877  wyoming  https://wyoming.craigslist.org  34990  2020.0     cadillac   
426878  wyoming  https://wyoming.craigslist.org  28990  2018.0        lexus   
426879  wyoming  https://wyoming.craigslist.org  30590  2019.0          bmw   

                           model condition    cylinders  ... size       type  \
426875         maxima s sedan 4d      good  6 cylinders  ...  NaN      sedan   
426876  s60 t5 momentum sedan 4d      good          NaN  ...  NaN      sedan   
426877          xt4 sport suv 4d      good          NaN  ...  NaN  hatchback   
426878           es 350 sedan 4d      good  6 cylinders  ...  NaN      sedan   
426879  4 series 430i gran coupe      good          NaN  ...  NaN      coupe   

       paint_color                                          image_url  \
426875         NaN  https://images.craigslist.org/00o0o_iiraFnHg8q...   
426876         red  https://images.craigslist.org/00x0x_15sbgnxCIS...   
426877       white  https://images.craigslist.org/00L0L_farM7bxnxR...   
426878      silver  https://images.craigslist.org/00z0z_bKnIVGLkDT...   
426879         NaN  https://images.craigslist.org/00Y0Y_lEUocjyRxa...   

                                              description county state  \
426875  Carvana is the safer way to buy a car During t...    NaN    wy   
426876  Carvana is the safer way to buy a car During t...    NaN    wy   
426877  Carvana is the safer way to buy a car During t...    NaN    wy   
426878  Carvana is the safer way to buy a car During t...    NaN    wy   
426879  Carvana is the safer way to buy a car During t...    NaN    wy   

              lat       long              posting_date  
426875  33.786500 -84.445400  2021-04-04T03:21:31-0600  
426876  33.786500 -84.445400  2021-04-04T03:21:29-0600  
426877  33.779214 -84.411811  2021-04-04T03:21:17-0600  
426878  33.786500 -84.445400  2021-04-04T03:21:11-0600  
426879  33.779214 -84.411811  2021-04-04T03:21:07-0600  

[5 rows x 26 columns]

In [30]:
# Affciher les lignes du df, aléatoirement
df.sample(n=10)

id                                                url  \
328635  7309700695  https://erie.craigslist.org/ctd/d/madison-2013...   
135621  7306631469  https://pullman.craigslist.org/ctd/d/spokane-2...   
246619  7316518253  https://lasvegas.craigslist.org/ctd/d/north-la...   
87276   7306129178  https://newhaven.craigslist.org/ctd/d/bridgepo...   
317855  7310930127  https://eugene.craigslist.org/ctd/d/albany-201...   
184575  7312473908  https://frederick.craigslist.org/cto/d/frederi...   
158838  7304154714  https://desmoines.craigslist.org/ctd/d/bellevu...   
152534  7309315230  https://indianapolis.craigslist.org/ctd/d/indi...   
377907  7315623210  https://lubbock.craigslist.org/ctd/d/lubbock-d...   
273264  7313436211  https://longisland.craigslist.org/ctd/d/freepo...   

                  region                           region_url  price    year  \
328635              erie          https://erie.craigslist.org  13995  2013.0   
135621  pullman / moscow       https://pullman.craigslist.org      0  2011.0   
246619         las vegas      https://lasvegas.craigslist.org  39999  2015.0   
87276          new haven      https://newhaven.craigslist.org   9300  2007.0   
317855            eugene        https://eugene.craigslist.org  24995  2017.0   
184575         frederick     https://frederick.craigslist.org   2900  2000.0   
158838        des moines     https://desmoines.craigslist.org  14950  2014.0   
152534      indianapolis  https://indianapolis.craigslist.org  16590  2016.0   
377907           lubbock       https://lubbock.craigslist.org   2500  2010.0   
273264       long island    https://longisland.craigslist.org  23495  2012.0   

       manufacturer                     model  condition    cylinders  ...  \
328635         ford                  explorer        NaN          NaN  ...   
135621          gmc               sierra 1500        NaN  8 cylinders  ...   
246619          gmc      sierra 1500 crew cab        NaN  8 cylinders  ...   
87276         lexus                        rx        NaN          NaN  ...   
317855          ram  pickup 1500 big horn 4x4        NaN          NaN  ...   
184575         jeep            grand cherokee  excellent  8 cylinders  ...   
158838        volvo                       s60       good          NaN  ...   
152534      hyundai       sonata eco sedan 4d       good          NaN  ...   
377907       toyota      sequoia sr5 platinum  excellent  8 cylinders  ...   
273264         jeep                  wrangler        NaN  6 cylinders  ...   

            size    type paint_color  \
328635       NaN     SUV      silver   
135621       NaN   truck        grey   
246619       NaN  pickup       black   
87276        NaN   wagon         NaN   
317855       NaN     NaN         NaN   
184575       NaN     SUV      silver   
158838       NaN   sedan        blue   
152534       NaN   sedan       white   
377907  mid-size     SUV       white   
273264       NaN     SUV       green   

                                                image_url  \
328635  https://images.craigslist.org/00N0N_22Ir6328PL...   
135621  https://images.craigslist.org/01515_JmiBqWU2vP...   
246619  https://images.craigslist.org/00R0R_i1jzIEILEt...   
87276   https://images.craigslist.org/00s0s_dJFB0oWG2F...   
317855  https://images.craigslist.org/00A0A_fqLJUUBWIf...   
184575  https://images.craigslist.org/00606_bhHbMuI1OY...   
158838  https://images.craigslist.org/00J0J_gTlvZlVuI6...   
152534  https://images.craigslist.org/01616_arIGe3CiOt...   
377907  https://images.craigslist.org/00d0d_4nJ6Tmnwic...   
273264  https://images.craigslist.org/01111_l30ENGpgsq...   

                                              description county state  \
328635  2013 Ford Explorer Limited 4WD     Offered by:...    NaN    pa   
135621  This 2011 GMC Sierra 1500 SLE has been SOLD bu...    NaN    id   
246619  2015 GMC Sierra 1500 Crew Cab Offered by: Rebe...    NaN    nv   
87276   2007 Lexus RX 350 AWD 4dr     Offered by: CT A...    Na

In [31]:
# Dimension du df c-a-d le nombre de lignes et colonnes
df.shape

(426880, 26)

In [32]:
# Nombre de colonnes dans df
df.columns

Index(['id', 'url', 'region', 'region_url', 'price', 'year', 'manufacturer',
       'model', 'condition', 'cylinders', 'fuel', 'odometer', 'title_status',
       'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color',
       'image_url', 'description', 'county', 'state', 'lat', 'long',
       'posting_date'],
      dtype='object')

In [33]:
# Le type de données de chaque colonne du df
df.dtypes

id                int64
url              object
region           object
region_url       object
price             int64
year            float64
manufacturer     object
model            object
condition        object
cylinders        object
fuel             object
odometer        float64
title_status     object
transmission     object
VIN              object
drive            object
size             object
type             object
paint_color      object
image_url        object
description      object
county          float64
state            object
lat             float64
long            float64
posting_date     object
dtype: object

La commande ci-dessous, nous me permet d'avoir des informations sur le dataset comme le type de données et le nombre total des valeurs non nulles dans chaque colonne.

Cela nous permet didentifier les valeurs manquantes et les types de données incorrectes 

In [34]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

In [35]:
#Statistiques descriptives pour les colonnes numériques (moyenne, écart-type, min, max, quartiles)
df.describe()

id         price           year      odometer  county  \
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05     0.0   
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04     NaN   
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05     NaN   
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00     NaN   
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04     NaN   
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04     NaN   
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05     NaN   
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07     NaN   

                 lat           long  
count  420331.000000  420331.000000  
mean       38.493940     -94.748599  
std         5.841533      18.365462  
min       -84.122245    -159.827728  
25%        34.601900    -111.939847  
50%        39.150100     -88.432600  
75%        42.398900     -80.832039  
max        82.390818     173.885502

In [36]:
# Statistiques descriptives pour des colonnes catégorielles (nombre de valeurs uniques, fréquence des plus courantes).
df.describe(include='object')

url    region  \
count                                              426880    426880   
unique                                             426880       404   
top     https://wyoming.craigslist.org/ctd/d/atlanta-2...  columbus   
freq                                                    1      3608   

                            region_url manufacturer   model condition  \
count                           426880       409234  421603    252776   
unique                             413           42   29667         6   
top     https://spokane.craigslist.org         ford   f-150      good   
freq                              2988        70985    8009    121456   

          cylinders    fuel title_status transmission                VIN  \
count        249202  423867       418638       424324             265838   
unique            8       5            6            3             118264   
top     6 cylinders     gas        clean    automatic  1FMJU1JT1HEA52352   
freq          94169  356209       405117       336524                261   

         drive       size    type paint_color  \
count   296313     120519  334022      296677   
unique       3          4      13          12   
top        4wd  full-size   sedan       white   
freq    131904      63465   87056       79285   

                                                image_url  \
count                                              426812   
unique                                             241899   
top     https://images.craigslist.org/00N0N_1xMPvfxRAI...   
freq                                                 7357   

                                              description   state  \
count                                              426810  426880   
unique                                             360911      51   
top     35 VEHICLES PRICED UNDER $3000!!!  BIG TIME! T...      ca   
freq                                                  231   50614   

                    posting_date  
count                     426812  
unique                    381536  
top     2021-04-23T22:13:05-0400  
freq                          12

## Nettoyage

Cette partie consiste à manipuler le dataset dans le but de le nettoyer et de le traiter.

Pour ce faire, nous allons procéder en gérant les valeurs manquantes, doublons, abbérantes et corriger si nécessaire les types de données des colonnes non conformes



### Gestion des valeurs manquantes

In [37]:
# Verifier la présence des valeurs manaquantes

valeurs_manquantes = df.isnull().sum()
print(valeurs_manquantes)

id                   0
url                  0
region               0
region_url           0
price                0
year              1205
manufacturer     17646
model             5277
condition       174104
cylinders       177678
fuel              3013
odometer          4400
title_status      8242
transmission      2556
VIN             161042
drive           130567
size            306361
type             92858
paint_color     130203
image_url           68
description         70
county          426880
state                0
lat               6549
long              6549
posting_date        68
dtype: int64


Pour gérer ces valeurs manquantes nous allons procéder comme suit : 
1. Selectionner les variables non pertinentes (qui ne sont assez décisif pour acheter une voiture) puis à les retirer du dataset.

Ces variables sont :

*County* qui n'a aucunes données, spécifie à renseigner le lieu où l'annonce a été postée.

*Size* qui contient beaucoup de valeurs manquantes d'ordre 300k, est une variable qu'on peut s'empasser lors des analyses car la variable *Type* donne l'information sur le design du véhicule.

*VIN* présente beaucoup de NA d'ordre 100k, n'est nécessaire pour l'analyse car elle symbolise le matricule de la voiture. Le matricule d'une voiture est une donnée qui peut être changeable par son nouveau propriétaire.

*Description*  n'a pas de pertinence car elle est similaire à page 'à propos' d'un site. Elle n'apporte pas d'information cruciale sur le véhicule.

*lat & long* represente les coordonnées géographiques de l'endroit où l'annonce a été faite.

*image_url* non nécessaire pour l'analyse, aussi les liens sont non fonctionnels qui peut être du au fait que l'annonce ait été retiré par l'achat du vehicule ou pas.

Enfin les variables *id, url, region, et region_url* certe n'ont aucune valeurs manquantes cependant elles servent à titre d'identifier ou de référence de l'annonce.



In [14]:
colonnes_a_supprimer = df[['VIN', 'description', 'county', 'id','lat', 'long', 
                           'size', 'url', 'region', 'region_url', 'image_url']]

In [15]:
colonnes_a_supprimer

VIN                                        description  \
0                     NaN                                                NaN   
1                     NaN                                                NaN   
2                     NaN                                                NaN   
3                     NaN                                                NaN   
4                     NaN                                                NaN   
...                   ...                                                ...   
426875  1N4AA6AV6KC367801  Carvana is the safer way to buy a car During t...   
426876  7JR102FKXLG042696  Carvana is the safer way to buy a car During t...   
426877  1GYFZFR46LF088296  Carvana is the safer way to buy a car During t...   
426878  58ABK1GG4JU103853  Carvana is the safer way to buy a car During t...   
426879  WBA4J1C58KBM14708  Carvana is the safer way to buy a car During t...   

        county          id        lat       long size  \
0          NaN  7222695916        NaN        NaN  NaN   
1          NaN  7218891961        NaN        NaN  NaN   
2          NaN  7221797935        NaN        NaN  NaN   
3          NaN  7222270760        NaN        NaN  NaN   
4          NaN  7210384030        NaN        NaN  NaN   
...        ...         ...        ...        ...  ...   
426875     NaN  7301591192  33.786500 -84.445400  NaN   
426876     NaN  7301591187  33.786500 -84.445400  NaN   
426877     NaN  7301591147  33.779214 -84.411811  NaN   
426878     NaN  7301591140  33.786500 -84.445400  NaN   
426879     NaN  7301591129  33.779214 -84.411811  NaN   

                                                      url  \
0       https://prescott.craigslist.org/cto/d/prescott...   
1       https://fayar.craigslist.org/ctd/d/bentonville...   
2       https://keys.craigslist.org/cto/d/summerland-k...   
3       https://worcester.craigslist.org/cto/d/west-br...   
4       https://greensboro.craigslist.org/cto/d/trinit...   
...                                                   ...   
426875  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

                        region                         region_url  \
0                     prescott    https://prescott.craigslist.org   
1                 fayetteville       https://fayar.craigslist.org   
2                 florida keys        https://keys.craigslist.org   
3       worcester / central MA   https://worcester.craigslist.org   
4                   greensboro  https://greensboro.craigslist.org   
...                        ...                                ...   
426875                 wyoming     https://wyoming.craigslist.org   
426876                 wyoming     https://wyoming.craigslist.org   
426877                 wyoming     https://wyoming.craigslist.org   
426878                 wyoming     https://wyoming.craigslist.org   
426879                 wyoming     https://wyoming.craigslist.org   

                                                image_url  
0                                                     NaN  
1                                                     NaN  
2                                                     NaN  
3                                                     NaN  
4                                                     NaN  
...                                                   ...  
426875  https://images.craigslist.org/00o0o_iiraFnHg8q...  
426876  https://images.craigslist.org/00x0x_15sbgnxCIS...  
426877  https://images.craigslist.org/00L0L_farM7bxnxR...  
426878  https://images.craigslist.org/00z0z_bKnIVGLkDT...  
426879  https://images.craigslist.org/00Y0Y_lEUocjyRxa...  

[426880 rows x 11 columns]

In [18]:
df.drop(colonnes_a_supprimer, axis=1, inplace=True)

In [19]:
df

price    year manufacturer                     model condition  \
0        6000     NaN          NaN                       NaN       NaN   
1       11900     NaN          NaN                       NaN       NaN   
2       21000     NaN          NaN                       NaN       NaN   
3        1500     NaN          NaN                       NaN       NaN   
4        4900     NaN          NaN                       NaN       NaN   
...       ...     ...          ...                       ...       ...   
426875  23590  2019.0       nissan         maxima s sedan 4d      good   
426876  30590  2020.0        volvo  s60 t5 momentum sedan 4d      good   
426877  34990  2020.0     cadillac          xt4 sport suv 4d      good   
426878  28990  2018.0        lexus           es 350 sedan 4d      good   
426879  30590  2019.0          bmw  4 series 430i gran coupe      good   

          cylinders    fuel  odometer title_status transmission drive  \
0               NaN     NaN       NaN          NaN          NaN   NaN   
1               NaN     NaN       NaN          NaN          NaN   NaN   
2               NaN     NaN       NaN          NaN          NaN   NaN   
3               NaN     NaN       NaN          NaN          NaN   NaN   
4               NaN     NaN       NaN          NaN          NaN   NaN   
...             ...     ...       ...          ...          ...   ...   
426875  6 cylinders     gas   32226.0        clean        other   fwd   
426876          NaN     gas   12029.0        clean        other   fwd   
426877          NaN  diesel    4174.0        clean        other   NaN   
426878  6 cylinders     gas   30112.0        clean        other   fwd   
426879          NaN     gas   22716.0        clean        other   rwd   

             type paint_color state              posting_date  
0             NaN         NaN    az                       NaN  
1             NaN         NaN    ar                       NaN  
2             NaN         NaN    fl                       NaN  
3             NaN         NaN    ma                       NaN  
4             NaN         NaN    nc                       NaN  
...           ...         ...   ...                       ...  
426875      sedan         NaN    wy  2021-04-04T03:21:31-0600  
426876      sedan         red    wy  2021-04-04T03:21:29-0600  
426877  hatchback       white    wy  2021-04-04T03:21:17-0600  
426878      sedan      silver    wy  2021-04-04T03:21:11-0600  
426879      coupe         NaN    wy  2021-04-04T03:21:07-0600  

[426880 rows x 15 columns]

####  Analyse des variables title_status et condition

Les variables title_status et condition donnent la même information sur l'état du véhicule. Cependant les valeurs de la colonne *title_status* sont plus réalistes que l'autre colonne.

En effet, *title_status* ces observations sont objectives tandis que celles de *condition* sont subjectives c'est à dire basées sur le point de vue du vendeur.  

In [40]:
df['title_status'].unique()

array([nan, 'clean', 'rebuilt', 'lien', 'salvage', 'missing',
       'parts only'], dtype=object)

In [41]:
df['condition'].unique()

array([nan, 'good', 'excellent', 'fair', 'like new', 'new', 'salvage'],
      dtype=object)

*Qu'est-ce qu'une Matrice de Contingence ?*

Une matrice de contingence (ou tableau croisé dynamique) est un tableau qui affiche la distribution conjointe de deux variables catégorielles (ou qualitatives). 

En d'autres termes, elle permet de comprendre comment les fréquences d'une catégorie d'une variable se répartissent selon les catégories d'une autre variable.

In [ ]:
print("\nMatrice de contingence entre title_status et condition:")
contingency_table = pd.crosstab(df['title_status'], df['condition'])
print(contingency_table)


Matrice de contingence entre title_status et condition:
condition     excellent  fair    good  like new   new  salvage
title_status                                                  
clean             91734  6156  118461     19870  1226      225
lien                586    47     389       265    25        3
missing              57   188     208         8     5       49
parts only           18    27      29         9     4       41
rebuilt            2823   142    1362       791    30       23
salvage             880   209     893       235    15      260


Le constat montre qu'il y a de forte corrélation, ce qui permet de faire de l'imputation conditionnelle afin de compléter les valeurs NAN de la colonne *condition*

Par exemple, nous constatons que:
- 91734 voitures sont à la fois *clean et excellent*
- 118 461 voitures sont à la fois "clean et good"

La plupart des voitures sont en *excellent, good* soit en statut *clean* au lieu de *salvage*




In [43]:
print("\nMatrice de contingence entre title_status et condition avec NaN:")
contingency_table = pd.crosstab(df['title_status'], df['condition'], dropna=False)
print(contingency_table)


Matrice de contingence entre title_status et condition avec NaN:
condition     excellent  fair    good  like new   new  salvage     NaN
title_status                                                          
clean             91734  6156  118461     19870  1226      225  167445
lien                586    47     389       265    25        3     107
missing              57   188     208         8     5       49     299
parts only           18    27      29         9     4       41      70
rebuilt            2823   142    1362       791    30       23    2048
salvage             880   209     893       235    15      260    1376
NaN                5369     0     114         0     0        0    2759


Etant donnée que ces deux variables sont corrélées, on peut utiliser l'approche d'imputer les valeurs d'une variable en fonction des valeurs de l'autre variable. Cela s'appelle *l'imputation conditionnelle*

In [44]:
# Imputer les valeurs NA de title_status par missing 
# qui renvoit unknown (non-renseigné)

df['title_status'] = df['title_status'].fillna('missing')

In [ ]:
#vérifier
df['title_status'].isna().sum()

np.int64(0)

**Explication du code**

Dans la logique, un véhicule présentant un état de statut 'salvage ou endommagé', nous permet de présumer que sa condition sera aussi en mauvaise état qui correspond dans la variable *condition* à 'fear' ou 'salvage'


On selectionne les lignes qui ont d'une part une valeur **'NA'** dans la colonne *condition* et d'autre part une valeur **'salvage'** dans la colonne *title_status*

Dans l'affichage, seule la colonne **condition** appraitra vu qu'elle est renseigné comme deuxième argument de la location (loc) qui sera affecté par les valeurs *salvage* dans les champs NA.

De même avec une valeur *part_only* de la colonne *title_status* va plus correspondre à une condition *salvage ou fear*

In [48]:

df.loc[(df['condition'].isnull()) & (df['title_status'] == 'salvage'), 'condition'] = 'salvage'


In [49]:
df.loc[(df['condition'].isnull()) & (df['title_status'] == 'parts only'), 'condition'] = 'salvage' # ou 'fair'

In [ ]:
print("\nMatrice de contingence entre title_status et condition avec NaN après traitement:")
contingency_table = pd.crosstab(df['title_status'], df['condition'], dropna=False)
print(contingency_table)


Matrice de contingence entre title_status et condition avec NaN:
condition     excellent  fair    good  like new   new  salvage     NaN
title_status                                                          
clean             91734  6156  118461     19870  1226      225  167445
lien                586    47     389       265    25        3     107
missing            5426   188     322         8     5       49    3058
parts only           18    27      29         9     4      111       0
rebuilt            2823   142    1362       791    30       23    2048
salvage             880   209     893       235    15     1636       0


In [51]:
#la méthode .mode() prend la valeur la plus fréquente
df['condition'] = df.groupby('title_status')['condition'].transform(
    lambda x: x.fillna(x.mode()[0] 
                       if not x.mode().empty else 'unknown condition'))


In [52]:
print("\nMatrice de contingence entre title_status et condition avec NaN après traitement:")
contingency_table = pd.crosstab(df['title_status'], df['condition'], dropna=False)
print(contingency_table)


Matrice de contingence entre title_status et condition avec NaN après traitement:
condition     excellent  fair    good  like new   new  salvage
title_status                                                  
clean             91734  6156  285906     19870  1226      225
lien                693    47     389       265    25        3
missing            8484   188     322         8     5       49
parts only           18    27      29         9     4      111
rebuilt            4871   142    1362       791    30       23
salvage             880   209     893       235    15     1636


### Correction des types de données

### Gestion des doublons

### Gestion des abérrants (outlet)

## Axes d'analyse

## Vectorisation